File: practice/array/max_diff.py
arr = [2,3,10,6,4,8,1]
"""
max_diff = 0
for i in range(len(arr)):
    for j in range(i+1,len(arr)):
        if arr[j]-arr[i]>max_diff and j>i:
            max_diff = arr[j]-arr[i]
print(max_diff)

"""

minVal = arr[0]
res = arr[1] - arr[0]

for j in range(1,len(arr)):
    res = max(arr[j]-minVal,res)
    minVal = min(arr[j],minVal)

print(res)

File: again_start/geeksforgeeks/array/max_diff.cpp
#include<bits/stdc++.h>

using namespace std;

int naive(int arr[],int n){
    int res = arr[1] - arr[0];
    for(int i=0;i<n;i++){
        for(int j=i+1;j<n;j++){
            if (arr[j]-arr[i]>=res){
                res = arr[j]-arr[i];
            }
        }
    } 
    return res;
}

int efficient(int arr[],int n){
    int minVal = arr[0];
    int res = arr[1]-arr[0];
    for(int i=1;i<n;i++){
        res = max(res,arr[i]-minVal);
        minVal = min(minVal,arr[i]);
    }
    return res;
}


int main(int argc, char const *argv[])
{
    int arr[]={2,3,10,6,4,8,1};
    int n = sizeof(arr)/sizeof(arr[0]);
    for(int i = 0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    cout<<efficient(arr,n);
    
    return 0;
}
